src/ml/algorithm_analysis.py

```diff
@@ -0,0 +1,131 @@
+import pandas as pd
+from typing import Iterable
+from pathlib import Path
+import os
+
+import matplotlib.pyplot as plt
+from sklearn.decomposition import PCA
+from sklearn.preprocessing import StandardScaler
+import seaborn as sns
+
+from sklearn.cluster import AgglomerativeClustering
+from scipy.cluster.hierarchy import dendrogram
+import numpy as np
+
+import argparse
+
+def is_file_empty(filepath):
+    return os.path.getsize(filepath)==0
+
+def summarize_networks(file_paths: Iterable[Path]) -> pd.DataFrame:
+
+    edge_tuples = []
+    
+    for file in file_paths:
+        
+        try:
+           
+            with open(file,'r') as f:
+                lines = [line[:3] for line in f.readlines()]
+            
+            line = []
+            for char in lines:
+                newChar = char.replace(' ','')
+                line.append(newChar)
+                
+            e = [''.join(sorted(ele)) for ele in line]
+
+            # for smaller labels (havent put it yet because of the tester files, but will be changed to this)
+            parts = file.split('/')
+            edge_tuples.append((parts[-2], e))
+  
+        except FileNotFoundError:
+            print(file, ' DOES NOT EXIST') # should hopefully not hit
+            continue
+        
+    edge_dataframes = []
+    
+    for tuple in edge_tuples:
+        col_label = str(tuple[0])
+
+        dataframe = pd.DataFrame(
+            {
+                col_label: 1,
+            }, index= tuple[1]
+        )
+        edge_dataframes.append(dataframe)
+
+    concated_df = pd.concat(edge_dataframes, axis= 1, join = 'outer')
+    concated_df = concated_df.fillna(0) 
+    concated_df = concated_df.astype('int64')
+   
+    return concated_df
+    
+def pca(dataframe: pd.DataFrame, output_png: str, output_file: str):
+
+    df = dataframe.reset_index(drop=True)
+    df = df.transpose() #based on the algortihms rather than the edges 
+    X = df.values
+
+    # standardize the df because if not in the same scale (rn it is 4,7)
+    scaler = StandardScaler()
+    scaler.fit(X) # calc mean and standard deviation 
+    X_scaled = scaler.transform(X)
+
+    # chosing the PCA
+    pca_2 = PCA(n_components=2)
+    pca_2.fit(X_scaled)
+    X_pca_2 = pca_2.transform(X_scaled)
+    variance = pca_2.explained_variance_ratio_ *100
+
+    # making the plot
+    plt.figure(figsize = (10,7))
```

Comment:
I get a warning here
```
UserWarning: Starting a Matplotlib GUI outside of the main thread will likely fail.
```

There was also a warning at the end of the run
```
QObject::~QObject: Timers cannot be stopped from another thread
```

Reply:
Based on what I found online, Matplotlib isn't thread safe.

Reply:
https://matplotlib.org/3.1.0/faq/howto_faq.html#working-with-threads

https://stackoverflow.com/questions/34764535/why-cant-matplotlib-plot-in-a-different-thread

Reply:
`matplotlib.use('Agg')`

maybe part of the solution https://stackoverflow.com/questions/41432010/strange-error-qobjectqobject-timers-cannot-be-stopped-from-another-thread

Reply:
/Users/nehatalluri/Desktop/jobs/research/spras/src/analysis/ml/ml.py:84: UserWarning: Starting a Matplotlib GUI outside of the main thread will likely fail.
  plt.figure(figsize = (10,7))
2023-02-22 15:35:13.107 python3.8[80211:4796193] WARNING: NSWindow drag regions should only be invalidated on the Main Thread! This will throw an exception in the future. Called from (
        0   AppKit                              0x00007ff8223cb0c3 -[NSWindow(NSWindow_Theme) _postWindowNeedsToResetDragMarginsUnlessPostingDisabled] + 352
        1   AppKit                              0x00007ff8223b7054 -[NSWindow _initContent:styleMask:backing:defer:contentView:] + 1288
        2   AppKit                              0x00007ff8223b6b46 -[NSWindow initWithContentRect:styleMask:backing:defer:] + 42
        3   _macosx.cpython-38-darwin.so        0x000000010b92519c -[Window initWithContentRect:styleMask:backing:defer:withManager:] + 76
        4   _macosx.cpython-38-darwin.so        0x000000010b928719 FigureManager_init + 281
        5   python3.8                           0x0000000106b48ddc wrap_init + 12
        6   python3.8                           0x0000000106ad2c63 wrapperdescr_call + 627
        7   python3.8                           0x0000000106ac5286 _PyObject_MakeTpCall + 214
        8   python3.8                           0x0000000106bfd855 call_function + 309
        9   python3.8                           0x0000000106bf59f2 _PyEval_EvalFrameDefault + 26930
        10  python3.8                           0x0000000106ac61d5 function_code_fastcall + 149
        11  python3.8                           0x0000000106ac4ffe _PyObject_FastCallDict + 94
        12  python3.8                           0x0000000106b48bb2 slot_tp_init + 178
        13  python3.8                           0x0000000106b543f8 type_call + 296
        14  python3.8                           0x0000000106ac5286 _PyObject_MakeTpCall + 214
        15  python3.8                           0x0000000106bfd855 call_function + 309
        16  python3.8                           0x0000000106bf59f2 _PyEval_EvalFrameDefault + 26930
        17  python3.8                           0x0000000106ac61d5 function_code_fastcall + 149
        18  python3.8                           0x0000000106ac9e4d method_vectorcall + 141
        19  python3.8                           0x0000000106bfd7bf call_function + 159
        20  python3.8                           0x0000000106bf59f2 _PyEval_EvalFrameDefault + 26930
        21  python3.8                           0x0000000106becfe6 _PyEval_EvalCodeWithName + 566
        22  python3.8                           0x0000000106ac6345 _PyFunction_Vectorcall + 213
        23  python3.8                           0x0000000106ac9f1a method_vectorcall + 346
        24  python3.8                           0x0000000106ac58b9 PyVectorcall_Call + 121
        25  python3.8                           0x0000000106bf446e _PyEval_EvalFrameDefault + 21422
        26  python3.8                           0x0000000106becfe6 _PyEval_EvalCodeWithName + 566
        27  python3.8                           0x0000000106ac6345 _PyFunction_Vectorcall + 213
        28  python3.8                           0x0000000106ac58b9 PyVectorcall_Call + 121
        29  python3.8                           0x0000000106bf446e _PyEval_EvalFrameDefault + 21422
        30  python3.8                           0x0000000106becfe6 _PyEval_EvalCodeWithName + 566
        31  python3.8                           0x0000000106ac6345 _PyFunction_Vectorcall + 213
        32  python3.8                           0x0000000106bfd7bf call_function + 159
        33  python3.8                           0x0000000106bf0cf7 _PyEval_EvalFrameDefault + 7223
        34  python3.8                           0x0000000106ac61d5 function_code_fastcall + 149
        35  python3.8                           0x0000000106bfd7bf call_function + 159
        36  python3.8                           0x0000000106bf59f2 _PyEval_EvalFrameDefault + 26930
        37  python3.8                           0x0000000106ac61d5 function_code_fastcall + 149
        38  python3.8                           0x0000000106bfd7bf call_function + 159
        39  python3.8                           0x0000000106bf4267 _PyEval_EvalFrameDefault + 20903
        40  python3.8                           0x0000000106ac61d5 function_code_fastcall + 149
        41  python3.8                           0x0000000106ac58b9 PyVectorcall_Call + 121
        42  python3.8                           0x0000000106bf446e _PyEval_EvalFrameDefault + 21422
        43  python3.8                           0x0000000106becfe6 _PyEval_EvalCodeWithName + 566
        44  python3.8                           0x0000000106ac6345 _PyFunction_Vectorcall + 213
        45  python3.8                           0x0000000106ac9f1a method_vectorcall + 346
        46  python3.8                           0x0000000106ac58b9 PyVectorcall_Call + 121
        47  python3.8                           0x0000000106bf446e _PyEval_EvalFrameDefault + 21422
        48  python3.8                           0x0000000106ac61d5 function_code_fastcall + 149
        49  python3.8                           0x0000000106bfd7bf call_function + 159
        50  python3.8                           0x0000000106bf459a _PyEval_EvalFrameDefault + 21722
        51  python3.8                           0x0000000106ac61d5 function_code_fastcall + 149
        52  python3.8                           0x0000000106ac58b9 PyVectorcall_Call + 121
        53  python3.8                           0x0000000106bf446e _PyEval_EvalFrameDefault + 21422
        54  python3.8                           0x0000000106ac61d5 function_code_fastcall + 149
        55  python3.8                           0x0000000106bfd7bf call_function + 159
        56  python3.8                           0x0000000106bf459a _PyEval_EvalFrameDefault + 21722
        57  python3.8                           0x0000000106ac61d5 function_code_fastcall + 149
        58  python3.8                           0x0000000106bfd7bf call_function + 159
        59  python3.8                           0x0000000106bf459a _PyEval_EvalFrameDefault + 21722
        60  python3.8                           0x0000000106ac61d5 function_code_fastcall + 149
        61  python3.8                           0x0000000106ac9f97 method_vectorcall + 471
        62  python3.8                           0x0000000106ac58b9 PyVectorcall_Call + 121
        63  python3.8                           0x0000000106ce2683 t_bootstrap + 195
        64  python3.8                           0x0000000106c72eb4 pythread_wrapper + 36
        65  libsystem_pthread.dylib             0x00007ff81f8a44e1 _pthread_start + 125
        66  libsystem_pthread.dylib             0x00007ff81f89ff6b thread_start + 15
)
/Users/nehatalluri/Desktop/jobs/research/spras/src/analysis/ml/ml.py:128: UserWarning: Starting a Matplotlib GUI outside of the main thread will likely fail.
  plt.figure(figsize=(10,7))

Reply:
https://stackoverflow.com/questions/63412583/nswindow-drag-regions-should-only-be-invalidated-on-the-main-thread-this-will-t